### ISSUE TITLE
[Form] Change the order of fields

### ISSUE THREAD

**Issue by tmartin** [symfony/symfony#5827]
Hi,

Since the children of a form are stored in an `array()` in the `FormBuilder`, the order of fields when rendering of a form in the view **using only `form_widget(form)`** is determined by the way you added your fields in your `FormType`.

``` php
<?php

public function buildForm(FormBuilderInterface $builder, array $options)
{
    $builder
        ->add('field1', 'text')
        ->add('field2', 'text');

    // and so on...
}
```

**The problem is :** when using an `EventSubscriber` (like `preSetData`) in order to add or some fields in your form, you cannot choose where you want to add these fields.

``` php
<?php

public function preSetData(DataEvent $event)
{
    $data = $event->getData();
    $form = $event->getForm();

    if (null !== $data) {
        $form
            ->add($this->factory->createNamed('field3', 'text'))
            ->add($this->factory->createNamed('field4', 'text'));

        // and so on
    }
}
```

This example will render the form in this order :
- field3
- field4
- field1
- field2

In my case it's really annoying because I can't modify the view. The redering of my forms is generic and done **only using `form_widget(form)`**.

So, is there any way to specify the order of the fields when using an `EventSubscriber` ?

**Comment by henrypenny** [comment]
It seems to me that form field ordering should be a function of the formView not the form itself.
Also, Form->createView() tends to dictate a one-one relationship between forms and form views, which I don't think is correct.

Perhaps `Form::createView()` could accept an argument that provides customisation for the form view.

`
$form->createView([
    'sort' => 'field2,field1',
    'sort-end' => 'field n-1, field n'
])
`

Any fields not included in the sort values could fall into place in their natural order.

This paradigm could be further extended to handle other common use cases:

`
$form->createView([
    'hidden' => 'field5'
])
`

Converts a given field to a hidden value to type="hidden".

`
$form->createView([
    'setRendered' => 'field6'
])
`

Does the same as `{% do form.field6.setRendered %}`

Using an array for this purpose would make it easy for form views to be configured via yaml.

`$form->createView(['formView' => 'userView'])`

or

`$form->createView(['formView' => 'superUserView'])`

All of this occurs after the form itself has been fully built so no issues with events.

**Comment by henrypenny** [comment]
Also, access to a form view's children is public. So you can sort them directly.

        uasort($view->children, function(FormView $a, FormView $b) {
            // do a comparison
        });

It there anything inherently wrong with doing this?

**Comment by juanmf** [comment]
@henrypenny I think you should not alter the state of shared objects. Ordering a form like this in client code might bring trouble.

**Comment by henrypenny** [comment]
How is a FormView a shared object? Its normally created just before its rendered then abandoned is it not?

**Comment by juanmf** [comment]
You got a point :)

**Comment by ericksho** [comment]
Will this be in core?

**Comment by Glideh** [comment]
What is the official feature/workaround about it ?
@AndrewCarterUK idea will work in most cases (adding all fields into the listener) but not so handy with multiple listeners, and the performance should be impacted anyway.

**Comment by kbond** [comment]
There has been some movement on #11241 lately

**Comment by behramcelen** [comment]
👍  for weight on core. probably weight will be more useful then after/before. Also symfony event subscribing works with weight logic.

**Comment by backbone87** [comment]
Maybe this should be closed since #11241 was too

**Comment by xabbuh** [comment]
closing here (see https://github.com/symfony/symfony/pull/11241#issuecomment-288462731)
